File: games/lynrummy/python/geometry_plan.py
"""
geometry_plan.py — the unified geometry post-pass for the
agent's primitive emission.

Verbs (`verbs.py`) emit a logical primitive sequence that's
geometry-agnostic. `plan_actions` walks the sequence and
injects pre-flight `move_stack` primitives anywhere a
primitive would otherwise produce a board where two stacks
are within PACK_GAP of each other.

The agent's invariant: after every primitive applies, no two
stacks overlap (with PACK_GAP padding — the human-feel
threshold, stricter than the referee's legal margin). A human
player relocates crowded stacks BEFORE building on them; the
agent matches by injecting MoveStacks at the points where the
next primitive would otherwise produce a too-close result.

Mirrors `Game.Agent.GeometryPlan` (Elm) verbatim.
"""

import geometry
import primitives
import strategy

PACK_GAP = 30


def plan_actions(board, actions):
    """Walk a primitive sequence; inject MoveStack pre-flights
    where the next primitive would land the post-board in a
    too-close state. Returns the augmented sequence."""
    out = []
    sim = list(board)
    for action in actions:
        emitted, sim = _plan_one(sim, action)
        out.extend(emitted)
    return out


def _plan_one(sim, action):
    """Plan one primitive. If post-board's NEW stacks are
    pack-gap-clear from PRE-EXISTING stacks, emit as is.
    Otherwise try a pre-flight MoveStack of the perturbed
    stack to a clear loc and re-emit.

    The diff-based check (new vs pre-existing) is the right
    shape: a split's two halves are inherently close (+8px
    sibling offset) but that's not a violation since they
    came from the same parent. The check only flags when a
    primitive's output is too close to a stack that was
    already on the board."""
    post = primitives.apply_locally(sim, action)
    if _is_clean_after_action(sim, post):
        return [action], post
    pre_flight = _pre_flight(sim, action)
    if pre_flight is not None:
        move_prim, new_action, new_post = pre_flight
        return [move_prim, new_action], new_post
    return [action], post


def _pre_flight(sim, action):
    """Compute a pre-flight move for a primitive whose
    post-board would overlap. Returns (move, new_action,
    post-state) or None if no helpful pre-flight exists for
    this primitive shape."""
    kind = action["action"]
    if kind == "split":
        return _pre_flight_split(sim, action)
    if kind == "merge_stack":
        return _pre_flight_merge_stack(sim, action)
    return None


def _pre_flight_split(sim, action):
    """Move the source stack to a pack-gap-cleared loc with
    room for its full size, then re-emit the split. The
    post-split spawn lands within the source's relocated
    footprint."""
    si = action["stack_index"]
    src = sim[si]
    source_size = len(src["board_cards"])
    others = [s for i, s in enumerate(sim) if i != si]
    new_loc = geometry.find_open_loc(others, card_count=source_size)
    if new_loc == src["loc"]:
        return None
    move = {"action": "move_stack",
            "stack_index": si, "new_loc": new_loc}
    after_move = primitives.apply_locally(sim, move)
    # The relocated source keeps its content; the index may
    # change. Find by content.
    src_content = primitives.cards_of(src)
    new_si = primitives.find_stack_index(after_move, src_content)
    new_split = {"action": "split",
                 "stack_index": new_si,
                 "card_index": action["card_index"]}
    after_split = primitives.apply_locally(after_move, new_split)
    return move, new_split, after_split


def _pre_flight_merge_stack(sim, action):
    """Move the merge target to a pack-gap-cleared loc that
    fits the augmented (source+target) stack, then re-emit
    the merge."""
    src_si = action["source_stack"]
    tgt_si = action["target_stack"]
    src = sim[src_si]
    tgt = sim[tgt_si]
    source_size = len(src["board_cards"])
    target_size = len(tgt["board_cards"])
    final_size = source_size + target_size
    others = [s for i, s in enumerate(sim) if i not in (src_si, tgt_si)]
    final_loc = geometry.find_open_loc(others, card_count=final_size)
    side = action.get("side", "right")
    if side == "left":
        target_loc = {
            "left": final_loc["left"] + source_size * geometry.CARD_PITCH,
            "top": final_loc["top"],
        }
    else:
        target_loc = final_loc
    if target_loc == tgt["loc"]:
        return None
    move = {"action": "move_stack",
            "stack_index": tgt_si, "new_loc": target_loc}
    after_move = primitives.apply_locally(sim, move)
    src_content = primitives.cards_of(src)
    tgt_content = primitives.cards_of(tgt)
    new_src_si = primitives.find_stack_index(after_move, src_content)
    new_tgt_si = primitives.find_stack_index(after_move, tgt_content)
    new_merge = {"action": "merge_stack",
                 "source_stack": new_src_si,
                 "target_stack": new_tgt_si,
                 "side": side}
    after_merge = primitives.apply_locally(after_move, new_merge)
    return move, new_merge, after_merge


def _is_clean_after_action(pre_board, post_board):
    """Diff-based pack-gap check: new stacks (in post-board
    but not pre-board) must be pack-gap-clear from pre-existing
    stacks (stacks that survived from pre-board to post-board).
    New-vs-new pairs (split siblings) are exempt — they're
    inherently close by the +8px split offset, but that's not
    a primitive emitting an overlap with the rest of the
    board, so we don't flag it.

    Out-of-bounds check applies to all stacks unconditionally.
    """
    pre_keys = {_stack_key(s) for s in pre_board}
    pre_existing = [s for s in post_board if _stack_key(s) in pre_keys]
    new_stacks = [s for s in post_board if _stack_key(s) not in pre_keys]

    for s in post_board:
        l, t, r, b = geometry.stack_rect(s)
        if (l < 0 or t < 0 or
                r > geometry.BOARD_MAX_WIDTH or
                b > geometry.BOARD_MAX_HEIGHT):
            return False

    for new in new_stacks:
        new_padded = geometry.pad_rect(
            geometry.stack_rect(new), PACK_GAP)
        for old in pre_existing:
            if geometry.rects_overlap(
                    new_padded, geometry.stack_rect(old)):
                return False
    return True


def _stack_key(s):
    """Identity for diffing pre/post boards. Stacks are
    identified by content + loc so that a stack moved to a
    new loc reads as 'new' and the original as 'removed'."""
    cards = tuple(
        (bc["card"]["value"], bc["card"]["suit"],
         bc["card"]["origin_deck"])
        for bc in s["board_cards"])
    return (s["loc"]["top"], s["loc"]["left"], cards)

File: games/lynrummy/ts/src/geometry_plan.ts
// geometry_plan.ts — TS port of python/geometry_plan.py.
//
// Verbs (verbs.ts) emit a logical primitive sequence that's
// geometry-agnostic. `planActions` walks the sequence and injects
// pre-flight `move_stack` primitives anywhere a primitive would
// otherwise produce a board where two stacks are within PACK_GAP of
// each other.
//
// Agent invariant: after every primitive applies, no two stacks
// overlap (with PACK_GAP padding — the human-feel threshold,
// stricter than the referee's legal margin). A human player
// relocates crowded stacks BEFORE building on them; the agent matches
// by injecting MoveStacks at the points where the next primitive
// would otherwise produce a too-close result.
//
// Interior splits are pre-flighted unconditionally (Steve, 2026-04-23):
// even when the immediate post-board doesn't yet overlap, an interior
// split's siblings will sit in tight quarters that downstream
// primitives can't safely build on. Pre-moving the donor to a
// 4-side-clear region is the human-feel default.

import type { Card } from "./rules/card.ts";
import {
  type BoardStack, type Loc,
  CARD_PITCH, BOARD_MAX_WIDTH, BOARD_MAX_HEIGHT,
  stackRect, padRect, rectsOverlap, findOpenLoc,
} from "./geometry.ts";
import {
  type Primitive, type SplitPrim, type MergeStackPrim, type MoveStackPrim,
  applyLocally, cardsOf, findStackIndex,
} from "./primitives.ts";

export const PACK_GAP = 30;

/** Walk a primitive sequence; inject MoveStack pre-flights where the
 *  next primitive would land the post-board in a too-close state.
 *  Returns the augmented sequence. */
export function planActions(
  board: readonly BoardStack[],
  actions: readonly Primitive[],
): readonly Primitive[] {
  const out: Primitive[] = [];
  let sim: readonly BoardStack[] = board;
  for (const action of actions) {
    const { emitted, post } = planOne(sim, action);
    out.push(...emitted);
    sim = post;
  }
  return out;
}

function planOne(
  sim: readonly BoardStack[],
  action: Primitive,
): { emitted: Primitive[]; post: readonly BoardStack[] } {
  if (isInteriorSplit(sim, action)) {
    const pf = preFlight(sim, action);
    if (pf !== null) {
      return { emitted: [pf.move, pf.newAction], post: pf.newPost };
    }
  }
  const post = applyLocally(sim, action);
  if (isCleanAfterAction(sim, post)) {
    return { emitted: [action], post };
  }
  const pf = preFlight(sim, action);
  if (pf !== null) {
    return { emitted: [pf.move, pf.newAction], post: pf.newPost };
  }
  return { emitted: [action], post };
}

function isInteriorSplit(sim: readonly BoardStack[], action: Primitive): boolean {
  if (action.action !== "split") return false;
  const n = sim[action.stackIndex]!.cards.length;
  if (n < 3) return false;
  const ci = action.cardIndex;
  // Mirror applySplit's left_count derivation.
  const leftCount = ci + 1 <= Math.floor(n / 2) ? ci + 1 : ci;
  return leftCount !== 1 && leftCount !== n - 1;
}

interface PreFlightResult {
  readonly move: MoveStackPrim;
  readonly newAction: Primitive;
  readonly newPost: readonly BoardStack[];
}

function preFlight(sim: readonly BoardStack[], action: Primitive): PreFlightResult | null {
  if (action.action === "split") return preFlightSplit(sim, action);
  if (action.action === "merge_stack") return preFlightMergeStack(sim, action);
  return null;
}

function preFlightSplit(
  sim: readonly BoardStack[],
  action: SplitPrim,
): PreFlightResult | null {
  const si = action.stackIndex;
  const src = sim[si]!;
  const sourceSize = src.cards.length;
  const others = sim.filter((_, i) => i !== si);
  const newLoc = findOpenLoc(others, sourceSize);
  if (newLoc.top === src.loc.top && newLoc.left === src.loc.left) return null;
  const move: MoveStackPrim = {
    action: "move_stack", stackIndex: si, newLoc,
  };
  const afterMove = applyLocally(sim, move);
  const newSi = findStackIndex(afterMove, cardsOf(src));
  const newSplit: SplitPrim = {
    action: "split", stackIndex: newSi, cardIndex: action.cardIndex,
  };
  const afterSplit = applyLocally(afterMove, newSplit);
  return { move, newAction: newSplit, newPost: afterSplit };
}

function preFlightMergeStack(
  sim: readonly BoardStack[],
  action: MergeStackPrim,
): PreFlightResult | null {
  const srcSi = action.sourceStack;
  const tgtSi = action.targetStack;
  const src = sim[srcSi]!;
  const tgt = sim[tgtSi]!;
  const sourceSize = src.cards.length;
  const targetSize = tgt.cards.length;
  const finalSize = sourceSize + targetSize;
  const others = sim.filter((_, i) => i !== srcSi && i !== tgtSi);
  const finalLoc = findOpenLoc(others, finalSize);
  let targetLoc: Loc;
  if (action.side === "left") {
    targetLoc = {
      left: finalLoc.left + sourceSize * CARD_PITCH,
      top: finalLoc.top,
    };
  } else {
    targetLoc = finalLoc;
  }
  if (targetLoc.top === tgt.loc.top && targetLoc.left === tgt.loc.left) return null;
  const move: MoveStackPrim = {
    action: "move_stack", stackIndex: tgtSi, newLoc: targetLoc,
  };
  const afterMove = applyLocally(sim, move);
  const newSrcSi = findStackIndex(afterMove, cardsOf(src));
  const newTgtSi = findStackIndex(afterMove, cardsOf(tgt));
  const newMerge: MergeStackPrim = {
    action: "merge_stack",
    sourceStack: newSrcSi,
    targetStack: newTgtSi,
    side: action.side,
  };
  const afterMerge = applyLocally(afterMove, newMerge);
  return { move, newAction: newMerge, newPost: afterMerge };
}

/** Diff-based pack-gap check: new stacks (in post-board but not
 *  pre-board) must be pack-gap-clear from pre-existing stacks. New-
 *  vs-new pairs (split siblings) are exempt — they're inherently
 *  close by the +8px split offset, but that's not a primitive emitting
 *  an overlap with the rest of the board.
 *
 *  Out-of-bounds check applies to all stacks unconditionally. */
function isCleanAfterAction(
  preBoard: readonly BoardStack[],
  postBoard: readonly BoardStack[],
): boolean {
  const preKeys = new Set(preBoard.map(stackKey));
  const preExisting: BoardStack[] = [];
  const newStacks: BoardStack[] = [];
  for (const s of postBoard) {
    if (preKeys.has(stackKey(s))) preExisting.push(s);
    else newStacks.push(s);
  }

  for (const s of postBoard) {
    const r = stackRect(s);
    if (r.left < 0 || r.top < 0
        || r.right > BOARD_MAX_WIDTH || r.bottom > BOARD_MAX_HEIGHT) {
      return false;
    }
  }

  for (const fresh of newStacks) {
    const newPadded = padRect(stackRect(fresh), PACK_GAP);
    for (const old of preExisting) {
      if (rectsOverlap(newPadded, stackRect(old))) return false;
    }
  }
  return true;
}

/** Identity for diffing pre/post boards: content + loc. A stack that
 *  moves reads as "new" and the original as "removed." */
function stackKey(s: BoardStack): string {
  const cards = s.cards.map(c => `${c[0]},${c[1]},${c[2]}`).join(";");
  return `${s.loc.top},${s.loc.left}|${cards}`;
}
